graphblas/matrix.py

```diff
@@ -664,6 +675,120 @@ def from_values(
+        new_mat = ffi_new("GrB_Matrix*")
+        rv = Matrix._from_obj(new_mat, new_dtype, nrows, ncols, name=name)
+        if new_dtype._is_udt:
+            dtype_name = "UDT"
+        else:
+            dtype_name = new_dtype.name
+        call(
+            f"GrB_Matrix_import_{dtype_name}",
+            [
+                _Pointer(rv),
+                new_dtype,
+                _as_scalar(nrows, _INDEX, is_cscalar=True),
+                _as_scalar(ncols, _INDEX, is_cscalar=True),
+                _CArray(indptr),
+                _CArray(indices),
+                _CArray(values, dtype=new_dtype),
+                _as_scalar(indptr.size, _INDEX, is_cscalar=True),
+                _as_scalar(indices.size, _INDEX, is_cscalar=True),
+                _as_scalar(values.shape[0], _INDEX, is_cscalar=True),
+                fmt,
+            ],
+        )
+        return rv
+
+    @classmethod
+    def from_csr(cls, indptr, col_indices, values, dtype=None, *, ncols=None, name=None):
+        return cls._from_csx(_CSR_FORMAT, indptr, col_indices, values, dtype, ncols, name)
+
+    @classmethod
+    def from_csc(cls, indptr, row_indices, values, dtype=None, *, nrows=None, name=None):
+        return cls._from_csx(_CSC_FORMAT, indptr, row_indices, values, dtype, nrows, name)
+
+    def _to_csx(self, fmt, dtype=None):
+        Ap_len = _scalar_index("Ap_len")
+        Ai_len = _scalar_index("Ai_len")
+        Ax_len = _scalar_index("Ax_len")
+        call(
+            "GrB_Matrix_exportSize",
+            [_Pointer(Ap_len), _Pointer(Ai_len), _Pointer(Ax_len), fmt, self],
+        )
+        Ap = _CArray(size=Ap_len.gb_obj[0], name="&Ap")
+        Ai = _CArray(size=Ai_len.gb_obj[0], name="&Ai")
+        Ax = _CArray(size=Ax_len.gb_obj[0], name="&Ax", dtype=self.dtype)
+        if self.dtype._is_udt:
+            dtype_name = "UDT"
+        else:
+            dtype_name = self.dtype.name
+        call(
+            f"GrB_Matrix_export_{dtype_name}",
+            [Ap, Ai, Ax, _Pointer(Ap_len), _Pointer(Ai_len), _Pointer(Ax_len), fmt, self],
+        )
+
+        # Unwrap objects with walrus operator and trim to proper size if necessary
+        if (Ap := Ap.array).size > (Ap_len := Ap_len.gb_obj[0]):  # pragma: no cover
+            Ap = Ap[:Ap_len]
+        if (Ai := Ai.array).size > (Ai_len := Ai_len.gb_obj[0]):  # pragma: no cover
+            Ai = Ai[:Ai_len]
+        if (Ax := Ax.array).shape[0] > (Ax_len := Ax_len.gb_obj[0]):  # pragma: no cover
+            Ax = Ax[:Ax_len]
+
+        # Alt w/o walrus
+        # Ap = Ap.array
+        # Ai = Ai.array
+        # Ax = Ax.array
+        # Ap_len = Ap_len.gb_obj[0]
+        # Ai_len = Ai_len.gb_obj[0]
+        # Ax_len = Ax_len.gb_obj[0]
+        # if Ap.size > Ap_len:  # pragma: no cover
+        #     Ap = Ap[:Ap_len]
+        # if Ai.size > Ai_len:  # pragma: no cover
+        #     Ai = Ai[:Ai_len]
+        # if Ax.size > Ax_len:  # pragma: no cover
+        #     Ax = Ax[:Ax_len]
```

Comment:
@jim22k since I've been flirting with walrus recently (https://github.com/python-graphblas/python-graphblas/pull/287), do you like with or w/o walrus?  So much shorter with walrus.

Reply:
I can get behind the walrus in this case. It does seem to make it more readable. I do like your comment about the walrus. That is probably something we should always do for walrus uses in the codebase. Just to tip off readers.

Reply:
Agreed and done. Walrus can be dense, so comments as in this case are nice.